### PR TITLE
Improve KV related expr types, add r/w locking

### PR DESCRIPTION
**Pull request by vicpatel**
Also extended filter for compose record values

Please take a look.

### CODE REVIEW

File: pkg/expr/expr_types.go [hunk 1]
```diff
@@ -463,12 +463,12 @@ func CastToUnsignedInteger(val interface{}) (out uint64, err error) {
 }
 
 func (t *KV) Has(k string) bool {
-	_, has := t.value[k]
-	return has
+	_, is := t.value[k]
```
Comment (darh) [inline pkg/expr/expr_types.go:466]: why change this?

File: pkg/expr/expr_types.go [hunk 2]
```diff
@@ -463,12 +463,12 @@ func CastToUnsignedInteger(val interface{}) (out uint64, err error) {
...
+	return is
 }
 
 func (t *KV) Select(k string) (TypedValue, error) {
-	if v, has := t.value[k]; has {
+	if v, is := t.value[k]; is {
```
Comment (darh) [inline pkg/expr/expr_types.go:471]: why change this?

**Review by darh — CHANGES_REQUESTED** [review]

File: pkg/expr/expr_types.go (hunk 1 above)
Comment (vicpatel) [reply pkg/expr/expr_types.go:466]: Reverting this change, can't recall the reason but must be changed to keep all bool receivers same throughout the file 🙈

**Comment by vicpatel** [comment]
Merged into `2022.9`

**Pull request closed without merging**